client/components/tinymce/plugins/media/plugin.jsx

```diff
@@ -234,32 +298,45 @@ function mediaButton( editor ) {
 				return;
 			}
 
-			// The `img` object can be a string or a DOM node. We need a
-			// normalized string to replace the post content markup
-			let imgString;
-			if ( img.outerHTML ) {
-				imgString = img.outerHTML;
+			let mediaString = '';
+
+			if ( mediaHasCaption ) {
+				if ( captionNode.outerHTML ) {
+					mediaString = captionNode.outerHTML;
+
+					captionNode.outerHTML = event.content;
+				} else {
+					mediaString = captionNode;
+				}
+			} else if ( img.outerHTML ) {
+				// The `img` object can be a string or a DOM node. We need a
+				// normalized string to replace the post content markup
+				mediaString = img.outerHTML;
 
 				// In visual editing mode, we apply the changes immediately by
 				// mutating the DOM node directly
 				img.outerHTML = event.content;
 			} else {
-				imgString = img;
+				mediaString = img;
 			}
 
 			// Replace the instance in the original post content
 			if ( content ) {
-				content = content.replace( imgString, event.content );
+				content = content.replace( mediaString, event.content );
 			}
 
 			// Not only should the content be replaced here, but also for every
 			// instance in the undo stack.
 			//
 			// See: https://github.com/tinymce/tinymce/blob/4.2.4/js/tinymce/classes/EditorUpload.js#L49-L53
 			editor.undoManager.data = editor.undoManager.data.map( function( level ) {
-				level.content = level.content.replace( imgString, event.content );
+				level.content = level.content.replace( mediaString, event.content );
 				return level;
 			} );
+
+			if ( numOfImagesToUpdate ) {
```

Comment:
We probably won't reach this case, but just a warning that negative numbers are truthy.